File: python/card-bot/bot.py
# Copyright 2017 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
from apiclient.discovery import build, build_from_document
from flask import Flask, render_template, request, json, make_response
from httplib2 import Http
from oauth2client.service_account import ServiceAccountCredentials

app = Flask(__name__)

INTERACTIVE_TEXT_BUTTON_ACTION = "doTextButtonAction"
INTERACTIVE_IMAGE_BUTTON_ACTION = "doImageButtonAction"
INTERACTIVE_BUTTON_PARAMETER_KEY = "param_key"
BOT_HEADER = 'Card Bot Python'

@app.route('/', methods=['POST'])
def home_post():
    """Respond to POST requests to this endpoint.

    All requests sent to this endpoint from Hangouts Chat are POST
    requests.
    """

    event_data = request.get_json()

    resp = None

    # If the bot is removed from the space, it doesn't post a message
    # to the space. Instead, log a message showing that the bot was removed.
    if event_data['type'] == 'REMOVED_FROM_SPACE':
        logging.info('Bot removed from  %s' % event_data['space']['displayName'])
        return 'OK'

    elif event_data['type']  == 'ADDED_TO_SPACE' and event_data['space']['type'] == 'ROOM':
        resp = { 'text': ('Thanks for adding me to {}!'
            .format(event_data['space']['name'])) }

    elif event_data['type']  == 'ADDED_TO_SPACE' and event_data['space']['type'] == 'DM':
        resp = { 'text': ('Thanks for adding me to a DM, {}!'
            .format(event_data['user']['displayName'])) }

    elif event_data['type'] == 'MESSAGE':
        resp = create_card_response(event_data['message']['text'])

    elif event_data['type'] == 'CARD_CLICKED':
        action_name = event_data['action']['actionMethodName']
        parameters = event_data['action']['parameters']
        resp = respond_to_interactive_card_click(action_name, parameters)

    space_name = event_data['space']['name']

    logging.info(resp)

    # Uncomment the following line for a synchronous response.
    #return json.jsonify(resp)

    # Asynchronous response version:
    thread_id = None
    if event_data['message']['thread'] != None:
        thread_id = event_data['message']['thread']

    # Need to return a response to avoid an error in the Flask app
    send_async_response(resp, space_name, thread_id)
    return 'OK'

@app.route('/', methods=['GET'])
def home_get():
    """Respond to GET requests to this endpoint.

    This function responds to requests with a simple HTML landing page for this
    App Engine instance.
    """

    return render_template('home.html')


def send_async_response(response, space_name, thread_id):
    """Sends a response back to the Hangouts Chat room asynchronously.

    Args:
      response: the response payload
      spaceName: The URL of the Hangouts Chat room

    """

    # The following two lines of code update the thread that raised the event.
    # Delete them if you want to send the message in a new thread.
    if thread_id != None:
        response['thread'] = thread_id
    ##################################

    scopes = ['https://www.googleapis.com/auth/chat.bot']
    credentials = ServiceAccountCredentials.from_json_keyfile_name(
        'service-acct.json', scopes)
    http_auth = credentials.authorize(Http())

    chat = build('chat', 'v1', http=http_auth)
    chat.spaces().messages().create(
        parent=space_name,
        body=response).execute()

def create_card_response(event_message):
    """Creates a card response based on the message sent in Hangouts Chat.

    See the reference for JSON keys and format for cards:
    https://developers.google.com/hangouts/chat/reference/message-formats/cards

    Args:
        eventMessage: the user's message to the bot

    """

    response = dict()
    cards = list()
    widgets = list()
    header = None

    words = event_message.lower().split()

    for word in words:

        if word == 'header':
            header = {
                'header': {
                    'title': BOT_HEADER,
                    'subtitle': 'Card header',
                    'imageUrl': 'https://goo.gl/5obRKj',
                    'imageStyle': 'IMAGE'
                }
            }

        elif word == 'textparagraph':
            widgets.append({
                'textParagraph' : {
                    'text': '<b>This</b> is a <i>text paragraph</i>.'
                }
            })

        elif word == 'keyvalue':
            widgets.append({
                'keyValue': {
                    'topLabel': 'KeyValue Widget',
                    'content': 'This is a KeyValue widget',
                    'bottomLabel': 'The bottom label',
                    'icon': 'STAR'
                }
            })

        elif word == 'interactivetextbutton':
            widgets.append({
                'buttons': [
                    {
                        'textButton': {
                            'text': 'INTERACTIVE BUTTON',
                            'onClick': {
                                'action': {
                                    'actionMethodName': INTERACTIVE_TEXT_BUTTON_ACTION,
                                    'parameters': [{
                                        'key': INTERACTIVE_BUTTON_PARAMETER_KEY,
                                        'value': event_message
                                    }]
                                }
                            }
                        }
                    }
                ]
            })

        elif word == 'interactiveimagebutton':
            widgets.append({
                'buttons': [
                    {
                        'imageButton': {
                            'icon': 'EVENT_SEAT',
                            'onClick': {
                                'action': {
                                    'actionMethodName': INTERACTIVE_IMAGE_BUTTON_ACTION,
                                    'parameters': [{
                                        'key': INTERACTIVE_BUTTON_PARAMETER_KEY,
                                        'value': event_message
                                    }]
                                }
                            }
                        }
                    }
                ]
            })

        elif word == 'textbutton':
            widgets.append({
                'buttons': [
                    {
                        'textButton': {
                            'text': 'TEXT BUTTON',
                            'onClick': {
                                'openLink': {
                                    'url': 'https://developers.google.com',
                                }
                            }
                        }
                    }
                ]
            })

        elif word == 'imagebutton':
            widgets.append({
                'buttons': [
                    {
                        'imageButton': {
                            'icon': 'EVENT_SEAT',
                            'onClick': {
                                'openLink': {
                                    'url': 'https://developers.google.com',
                                }
                            }
                        }
                    }
                ]
            })

        elif word == 'image':
            widgets.append({
                'image': {
                    'imageUrl': 'https://goo.gl/Bpa3Y5',
                    'onClick': {
                        'openLink': {
                            'url': 'https://developers.google.com'
                        }
                    }
                }
            })



    if header != None:
        cards.append(header)

    cards.append({ 'sections': [{ 'widgets': widgets }]})
    response['cards'] = cards

    return response

def respond_to_interactive_card_click(action_name, custom_params):
    """Creates a response for when the user clicks on an interactive card.

    See the guide for creating interactive cards
    https://developers.google.com/hangouts/chat/how-tos/cards-onclick

    Args:
        action_name: the name of the custom action defined in the original bot response
        custom_params: the parameters defined in the original bot response

    """
    message = 'You clicked {}'.format(
        'a text button' if action_name == INTERACTIVE_TEXT_BUTTON_ACTION
            else 'an image button')

    original_message = ""

    if custom_params[0]['key'] == INTERACTIVE_BUTTON_PARAMETER_KEY:
        original_message = custom_params[0]['value']
    else:
        original_message = '<i>Cannot determine original message</i>'

    # If you want to respond to the same room but with a new message,
    # change the following value to NEW_MESSAGE.
    action_response = 'UPDATE_MESSAGE'

    return {
        'actionResponse': {
            'type': action_response
        },
        'cards': [
            {
                'header': {
                    'title': BOT_HEADER,
                    'subtitle': 'Interactive card clicked',
                    'imageUrl': 'https://goo.gl/5obRKj',
                    'imageStyle': 'IMAGE'
                }
            },
            {
                'sections': [
                    {
                        'widgets': [
                            {
                                'textParagraph': {
                                    'text': message
                                }
                            },
                            {
                                'keyValue': {
                                    'topLabel': 'Original message',
                                    'content': original_message
                                }
                            }
                        ]
                    }
                ]
            }
        ]
    }

File: java/card-bot/src/main/java/com/google/hangouts/chat/card/Bot.java
/*
 * Copyright 2017 Google Inc.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.hangouts.chat.card;

// [START card-bot]
import com.fasterxml.jackson.databind.JsonNode;
import com.google.api.services.chat.v1.model.*;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.*;
import java.util.logging.Logger;
import java.util.stream.Stream;

@SpringBootApplication
@RestController
public class Bot {
    private static final Logger logger = Logger.getLogger(Bot.class.getName());

    private static final String INTERACTIVE_TEXT_BUTTON_ACTION = "doTextButtonAction";
    private static final String INTERACTIVE_IMAGE_BUTTON_ACTION = "doImageButtonAction";
    private static final String INTERACTIVE_BUTTON_KEY = "originalMessage";
    private static final String HEADER_IMAGE = "https://goo.gl/5obRKj";
    private static final String BOT_NAME = "Card Bot Java";
    private static final String REDIRECT_URL = "https://goo.gl/kwhSNz";

    public static void main(String[] args) {
        SpringApplication.run(Bot.class, args);
    }

    /**
     * Handles a GET request to the /bot endpoint.
     *
     * @param event Event from chat.
     * @return Message
     */
    @PostMapping("/")
    @ResponseBody
    public Message onEvent(@RequestBody JsonNode event) {
        Message reply = new Message();
        String eventType = event.get("type").asText();
        switch (eventType) {
            case "ADDED_TO_SPACE":
                JsonNode space = event.get("space");
                String spaceType = space.get("type").asText();
                if ("ROOM".equals(spaceType)) {
                    String displayName = space.get("displayName").asText();
                    String replyText = String.format("Thanks for adding me to %s", displayName);
                    reply.setText(replyText);
                } else {
                    String displayName = event.get("user").get("displayName").asText();
                    String replyText = String.format("Thanks for adding me to a DM, %s!", displayName);
                    reply.setText(replyText);
                }
                break;
            case "MESSAGE":
                Card card = createCardResponse(event.get("message").get("text").asText());
                reply.setCards(Collections.singletonList(card));
                break;
            case "CARD_CLICKED":
                // Get the custom action name and custom parameter value out of the event object.
                String actionName = event.get("action").get("actionMethodName").asText();
                String customParameterValue = event.get("action").get("parameters").get(0).get("value").asText();
                card = respondToInteractiveCardClick(actionName, customParameterValue);
                reply.setCards(Collections.singletonList(card));
                break;
            case "REMOVED_FROM_SPACE":
                String name = event.get("space").get("name").asText();
                logger.info(String.format("Bot removed from %s", name));
                break;
            default:
                reply.setText("Cannot determine event type");
                break;
        }
        return reply;
    }


    /**
     * Creates a card-formatted response based on the message sent in Hangouts Chat.
     *
     * See the reference for JSON keys and format for cards:
     * https://developers.google.com/hangouts/chat/reference/message-formats/cards
     *
     * @param message the event object sent from Hangouts Chat
     * @return a card responseStr as a JSON object node
     */
    private Card createCardResponse(String message) {
        Card card = new Card();
        List<WidgetMarkup> widgets = new ArrayList<>();
        List<ActionParameter> customParameters = Collections.singletonList(
                new ActionParameter().setKey(INTERACTIVE_BUTTON_KEY).setValue(message)
        );

        // Parse the message and add widgets to the responseStr in the order that
        // they were requested in the message.
        Stream.of(message.split(" ")).forEach((s -> {
            if (s.contains("header")) {
                CardHeader header = new CardHeader()
                        .setTitle(BOT_NAME)
                        .setSubtitle("Card header")
                        .setImageUrl(HEADER_IMAGE)
                        .setImageStyle("IMAGE");
                card.setHeader(header);
            } else if (s.contains("textparagraph")) {
                TextParagraph widget = new TextParagraph().setText("<b>This</b> is a <i>text paragraph</i>.");
                widgets.add(new WidgetMarkup().setTextParagraph(widget));
            } else if (s.contains("keyvalue")) {
                KeyValue widget = new KeyValue()
                        .setTopLabel("KeyValue widget")
                        .setContent("This is a KeyValue widget")
                        .setBottomLabel("The bottom label")
                        .setIcon("STAR");
                widgets.add(new WidgetMarkup().setKeyValue(widget));
            } else if (s.contains("interactivetextbutton")) {
                FormAction action = new FormAction()
                        .setActionMethodName(INTERACTIVE_TEXT_BUTTON_ACTION)
                        .setParameters(customParameters);
                OnClick onClick = new OnClick().setAction(action);
                TextButton button = new TextButton()
                        .setText("INTERACTIVE BUTTON")
                        .setOnClick(onClick);
                Button widget = new Button().setTextButton(button);
                widgets.add(new WidgetMarkup().setButtons(Collections.singletonList((widget))));
            } else if (s.contains("interactiveimagebutton")) {
                FormAction action = new FormAction()
                        .setActionMethodName(INTERACTIVE_IMAGE_BUTTON_ACTION)
                        .setParameters(customParameters);
                OnClick onClick = new OnClick().setAction(action);
                ImageButton button = new ImageButton()
                        .setIcon("EVENT_SEAT")
                        .setOnClick(onClick);
                Button widget = new Button().setImageButton(button);
                widgets.add(new WidgetMarkup().setButtons(Collections.singletonList((widget))));
            } else if (s.contains("textbutton")) {
                OpenLink openLink = new OpenLink().setUrl(REDIRECT_URL);
                OnClick onClick = new OnClick().setOpenLink(openLink);
                TextButton button = new TextButton()
                        .setText("TEXT BUTTON")
                        .setOnClick(onClick);
                Button widget = new Button().setTextButton(button);
                widgets.add(new WidgetMarkup().setButtons(Collections.singletonList((widget))));
            } else if (s.contains("imagebutton")) {
                OpenLink openLink = new OpenLink().setUrl(REDIRECT_URL);
                OnClick onClick = new OnClick().setOpenLink(openLink);
                ImageButton button = new ImageButton()
                        .setIcon("EVENT_SEAT")
                        .setOnClick(onClick);
                Button widget = new Button().setImageButton(button);
                widgets.add(new WidgetMarkup().setButtons(Collections.singletonList((widget))));
            } else if (s.contains("image")) {
                Image widget = new Image().setImageUrl("https://goo.gl/Bpa3Y5");
                widgets.add(new WidgetMarkup().setImage(widget));
            }
        }));

        Section section = new Section()
                .setWidgets(widgets);
        card.setSections(Collections.singletonList(section));
        return card;
    }

    /**
     * Creates a responseStr for when the user clicks on an interactive card.
     *
     * @param actionName name of action invoked
     * @param customParameterValue custom payload from event
     * @return a response card
     */
    private Card respondToInteractiveCardClick(String actionName, String customParameterValue) {
        // Determine which button the user clicked.
        String message = String.format("You clicked <u>%s</u>. <br> Your original message was \"%s\".",
                actionName.equals(INTERACTIVE_TEXT_BUTTON_ACTION) ? "a TextButton" : "an ImageButton",
                customParameterValue);

        Card card = new Card();
        CardHeader header = new CardHeader();
        header.setTitle(BOT_NAME)
                .setSubtitle("Interactive card click")
                .setImageUrl(HEADER_IMAGE)
                .setImageStyle("IMAGE");
        card.setHeader(header);
        TextParagraph text = new TextParagraph().setText(message);
        List<WidgetMarkup> widgets = Collections.singletonList
                (new WidgetMarkup().setTextParagraph(text));
        Section section = new Section()
                .setWidgets(widgets);
        card.setSections(Collections.singletonList(section));
        return card;
    }
}
// [END card-bot]
